File: internal/domain/collection.go
package domain

import (
	"go.mongodb.org/mongo-driver/bson/primitive"
)

type FieldType string

const (
	// 1. Basic Input Components
	TypeTextInput    FieldType = "text"
	TypeTextArea     FieldType = "textarea"
	TypeNumberInput  FieldType = "number"
	TypeEmailInput   FieldType = "email"
	TypePasswordInput FieldType = "password"
	TypePhoneInput    FieldType = "phone"
	TypeURLInput      FieldType = "url"

	// 2. Selection Components
	TypeSelect      FieldType = "select"
	TypeRadio       FieldType = "radio"
	TypeCheckbox    FieldType = "checkbox"
	TypeMultiSelect FieldType = "multiselect"

	// 3. Date & Time Components
	TypeDatePicker     FieldType = "datepicker"
	TypeTimePicker     FieldType = "timepicker"
	TypeDateTimePicker FieldType = "datetimepicker"
	TypeDateRange      FieldType = "daterange"

	// 4. File & Media Components
	TypeFileUpload  FieldType = "file"
	TypeImageUpload FieldType = "image"

	// 5. Advanced Input Components
	TypeBool         FieldType = "bool" // Simple boolean
	TypeToggleField  FieldType = "toggle"
	TypeSliderField  FieldType = "slider"
	TypeRatingField  FieldType = "rating"
	TypeColorPickerField FieldType = "colorpicker"

	// 6. Layout & Structural Components
	TypeSection    FieldType = "section"
	TypeFieldGroup FieldType = "group"
	TypeTabs       FieldType = "tabs"
	TypeGrid       FieldType = "grid"

	// 8. Repeater / Dynamic List
	TypeRepeater FieldType = "repeater"

	// 9. Relational Components
	TypeRelation     FieldType = "relation"
	TypeAutocomplete FieldType = "autocomplete"

	// 10. Action Components
	TypeSubmitButton FieldType = "submit"
	TypeResetButton  FieldType = "reset"
	TypeCustomAction FieldType = "action"
)

type ValidationRules struct {
	Min       *float64 `bson:"min,omitempty" json:"min,omitempty"`
	Max       *float64 `bson:"max,omitempty" json:"max,omitempty"`
	MinLength *int     `bson:"min_length,omitempty" json:"min_length,omitempty"`
	MaxLength *int     `bson:"max_length,omitempty" json:"max_length,omitempty"`
	Regex     string   `bson:"regex,omitempty" json:"regex,omitempty"`
}

type ArrayConfig struct {
	MinItems    *int `bson:"min_items,omitempty" json:"min_items,omitempty"`
	MaxItems    *int `bson:"max_items,omitempty" json:"max_items,omitempty"`
	UniqueItems bool `bson:"unique_items" json:"unique_items"`
}

type ConditionalLogic struct {
	ShowIf *Condition `bson:"show_if,omitempty" json:"show_if,omitempty"`
}

type Condition struct {
	Field  string      `bson:"field" json:"field"`
	Equals interface{} `bson:"equals" json:"equals"`
}

type Option struct {
	Label string      `bson:"label" json:"label"`
	Value interface{} `bson:"value" json:"value"`
}

type RelationConfig struct {
	Collection string `bson:"collection" json:"collection"`
	Field      string `bson:"field" json:"field"` // UI display field
}

type Field struct {
	Key          string            `bson:"key" json:"key" validate:"required"`
	Label        string            `bson:"label" json:"label"`
	Type         FieldType         `bson:"type" json:"type" validate:"required"`
	Placeholder  string            `bson:"placeholder,omitempty" json:"placeholder,omitempty"`
	DefaultValue interface{}       `bson:"default_value,omitempty" json:"default_value,omitempty"`
	Required     bool              `bson:"required" json:"required"`
	Unique       bool              `bson:"unique" json:"unique"`
	Description  string            `bson:"description,omitempty" json:"description,omitempty"`
	Options      []Option          `bson:"options,omitempty" json:"options,omitempty"`
	Validation   *ValidationRules  `bson:"validation,omitempty" json:"validation,omitempty"`
	Logic        *ConditionalLogic `bson:"logic,omitempty" json:"logic,omitempty"`
	Fields       []Field           `bson:"fields,omitempty" json:"fields,omitempty"`

	// Relational & Advanced Logic
	Relation   *RelationConfig `bson:"relation,omitempty" json:"relation,omitempty"`
	ComputedBy string          `bson:"computed_by,omitempty" json:"computed_by,omitempty"`

	// Array Configuration
	IsArray     bool         `bson:"is_array" json:"is_array"`
	ArrayConfig *ArrayConfig `bson:"array_config,omitempty" json:"array_config,omitempty"`

	// Additional dynamic properties for specific components
	Props map[string]interface{} `bson:"props,omitempty" json:"props,omitempty"`

	// Query engine: include field in full-text-style search across records
	Searchable bool `bson:"searchable,omitempty" json:"searchable,omitempty"`
	// Strip from API responses for anonymous access to public collections
	Internal bool `bson:"internal,omitempty" json:"internal,omitempty"`
}

type CRUDPolicy struct {
	Create []string `bson:"create" json:"create"`
	Read   []string `bson:"read" json:"read"`
	Update []string `bson:"update" json:"update"`
	Delete []string `bson:"delete" json:"delete"`
}

type AccessControl struct {
	IsPublic     bool       `bson:"is_public" json:"is_public"`
	AllowedRoles []string   `bson:"allowed_roles,omitempty" json:"allowed_roles,omitempty"`
	CRUDPolicy   CRUDPolicy `bson:"crud_policy,omitempty" json:"crud_policy,omitempty"`
}

type Collection struct {
	ID          primitive.ObjectID `bson:"_id,omitempty" json:"id,omitempty"`
	Name        string             `bson:"name" json:"name" validate:"required"`
	Slug        string             `bson:"slug" json:"slug" validate:"required"`
	Fields      []Field            `bson:"fields" json:"fields" validate:"required,dive"`
	Description string             `bson:"description" json:"description"`
	Access      *AccessControl     `bson:"access,omitempty" json:"access,omitempty"`
}

File: Panel/src/lib/types/collection.ts
// ============================================================
// TypeScript interfaces mirroring Go domain models
// ============================================================

export type FieldType =
	// Basic Input
	| 'text'
	| 'textarea'
	| 'number'
	| 'email'
	| 'password'
	| 'phone'
	| 'url'
	// Selection
	| 'select'
	| 'radio'
	| 'checkbox'
	| 'multiselect'
	// Date & Time
	| 'datepicker'
	| 'timepicker'
	| 'datetimepicker'
	// File & Media
	| 'file'
	| 'image'
	// Advanced Input
	| 'bool'
	| 'toggle'
	| 'slider'
	| 'rating'
	| 'colorpicker'
	// Layout & Structure
	| 'section'
	| 'group'
	| 'tabs'
	| 'grid'
	// Repeater
	| 'repeater'
	// Relational
	| 'relation'
	| 'autocomplete'
	// Actions
	| 'submit'
	| 'reset'
	| 'action';

export interface ValidationRules {
	min?: number;
	max?: number;
	min_length?: number;
	max_length?: number;
	regex?: string;
}

export interface ArrayConfig {
	min_items?: number;
	max_items?: number;
	unique_items: boolean;
}

export interface Condition {
	field: string;
	equals: unknown;
}

export interface ConditionalLogic {
	show_if?: Condition;
}

export interface Option {
	label: string;
	value: unknown;
}

export interface RelationConfig {
	collection: string;
	field: string;
}

export interface Field {
	key: string;
	label: string;
	type: FieldType;
	placeholder?: string;
	default_value?: unknown;
	required: boolean;
	unique: boolean;
	description?: string;
	options?: Option[];
	validation?: ValidationRules;
	logic?: ConditionalLogic;
	fields?: Field[]; // recursive: group / repeater sub-fields
	relation?: RelationConfig;
	computed_by?: string;

	// Array Configuration
	is_array: boolean;
	array_config?: ArrayConfig;

	props?: Record<string, unknown>;
}

export interface CRUDPolicy {
	create: string[];
	read: string[];
	update: string[];
	delete: string[];
}

export interface AccessControl {
	is_public: boolean;
	allowed_roles?: string[];
	crud_policy?: CRUDPolicy;
}

export interface Collection {
	id?: string;
	name: string;
	slug: string;
	description: string;
	fields: Field[];
	access?: AccessControl;
}

// Record is a dynamic map from the content API
export type ContentRecord = Record<string, unknown> & { _id?: string; id?: string };

// Upload API response
export interface UploadResponse {
	path: string;
}

// ── Field type metadata for the UI ───────────────────────────

export interface FieldTypeGroup {
	label: string;
	types: { value: FieldType; label: string; icon: string }[];
}

export const FIELD_TYPE_GROUPS: FieldTypeGroup[] = [
	{
		label: 'Basic Input',
		types: [
			{ value: 'text', label: 'Text', icon: 'T' },
			{ value: 'textarea', label: 'Textarea', icon: '¶' },
			{ value: 'number', label: 'Number', icon: '#' },
			{ value: 'email', label: 'Email', icon: '@' },
			{ value: 'password', label: 'Password', icon: '🔒' },
			{ value: 'phone', label: 'Phone', icon: '☎' },
			{ value: 'url', label: 'URL', icon: '🔗' }
		]
	},
	{
		label: 'Selection',
		types: [
			{ value: 'select', label: 'Select', icon: '▾' },
			{ value: 'radio', label: 'Radio', icon: '◎' },
			{ value: 'checkbox', label: 'Checkbox', icon: '☑' },
			{ value: 'multiselect', label: 'Multi-Select', icon: '☰' }
		]
	},
	{
		label: 'Date & Time',
		types: [
			{ value: 'datepicker', label: 'Date', icon: '📅' },
			{ value: 'timepicker', label: 'Time', icon: '🕐' },
			{ value: 'datetimepicker', label: 'DateTime', icon: '📆' }
		]
	},
	{
		label: 'File & Media',
		types: [
			{ value: 'file', label: 'File', icon: '📄' },
			{ value: 'image', label: 'Image', icon: '🖼' }
		]
	},
	{
		label: 'Advanced',
		types: [
			{ value: 'bool', label: 'Boolean', icon: '⊙' },
			{ value: 'toggle', label: 'Toggle', icon: '⏻' },
			{ value: 'slider', label: 'Slider', icon: '⎓' },
			{ value: 'rating', label: 'Rating', icon: '★' },
			{ value: 'colorpicker', label: 'Color', icon: '🎨' }
		]
	},
	{
		label: 'Layout',
		types: [
			{ value: 'section', label: 'Section', icon: '▬' },
			{ value: 'group', label: 'Group', icon: '⊞' },
			{ value: 'tabs', label: 'Tabs', icon: '⊟' },
			{ value: 'grid', label: 'Grid', icon: '⊞' },
			{ value: 'repeater', label: 'Repeater', icon: '⟳' }
		]
	},
	{
		label: 'Relational',
		types: [
			{ value: 'relation', label: 'Relation', icon: '↔' },
			{ value: 'autocomplete', label: 'Autocomplete', icon: '⌕' }
		]
	}
];

export function getFieldTypeLabel(type: FieldType): string {
	for (const group of FIELD_TYPE_GROUPS) {
		const found = group.types.find((t) => t.value === type);
		if (found) return found.label;
	}
	return type;
}

export function isStructuralType(type: FieldType): boolean {
	return ['group', 'repeater', 'tabs', 'grid', 'section'].includes(type);
}

export function hasOptions(type: FieldType): boolean {
	return ['select', 'radio', 'multiselect', 'checkbox'].includes(type);
}

export function isFileType(type: FieldType): boolean {
	return ['file', 'image'].includes(type);
}
